typescript/src/redemption.ts

```diff
@@ -0,0 +1,291 @@
+ */
+async function prepareRedemptionRequests(
+  bridge: Bridge,
+  walletPrivateKey: string,
+  redeemerAddresses: string[]
+): Promise<RedemptionRequest[]> {
+  const walletKeyRing = createKeyRing(walletPrivateKey)
+  const walletPublicKey = walletKeyRing.getPublicKey().toString("hex")
+
+  const walletPubKeyHash = `0x${hash160
+    .digest(Buffer.from(walletPublicKey, "hex"))
+    .toString("hex")}`
+
+  const redemptionRequests: RedemptionRequest[] = []
+
+  for (const redeemerAddress of redeemerAddresses) {
+    const redeemerOutputScript = deriveOutputScript(redeemerAddress)
+
+    const pendingRedemption = await bridge.pendingRedemptions(
+      walletPubKeyHash,
+      redeemerOutputScript
+    )
+
+    if (pendingRedemption.requestedAt == 0) {
+      // The requested redemption does not exist among `pendingRedemptions`
+      // in the Bridge.
+      throw new Error(
+        "Provided redeemer address and wallet public key do not identify a pending redemption"
+      )
+    }
+
+    // TODO: Use the value of fee that was set in the Bridge (max fee) as the
+    // `txFee` for now.
+    // In the future allow the caller to propose the value of transaction fee.
+    // If the proposed transaction fee is smaller than the sum of fee shares from
+    // all the outputs then use the proposed fee and add the difference to outputs
+    // proportionally.
+
+    // Redemption exists in the Bridge. Add it to the list.
+    redemptionRequests.push({
+      redeemerAddress: redeemerAddress,
+      requestedAmount: pendingRedemption.requestedAmount,
+      treasuryFee: pendingRedemption.treasuryFee,
+      txFee: pendingRedemption.txFee,
+      requestedAt: pendingRedemption.requestedAt,
+    })
+  }
+
+  return redemptionRequests
+}
+
+/**
+ * Derives a length prefixed output script based on the provided address (P2PKH,
+ * P2WPKH, P2SH or P2WSH).
+ * @param address - Bitcoin address from which the output script will be build
+ * @returns The 0x-prefixed output script as a string.
+ */
+function deriveOutputScript(address: string): string {
+  const rawOutputScript = bcoin.Script.fromAddress(address).toRaw()
+
+  return `0x${Buffer.concat([
+    Buffer.from([rawOutputScript.length]),
+    rawOutputScript,
+  ]).toString("hex")}`
+}
+
+/**
+ * Creates a Bitcoin redemption transaction.
+ * The transaction will have a single input (main UTXO of the wallet making
+ * the redemption), an output for each redemption request provided, and a change
+ * output if the redemption requests do not consume the entire amount of the
+ * single input.
+ * @dev The caller is responsible for ensuring the redemption request list is
+ *      correctly formed:
+ *        - there is at least one redemption
+ *        - the `requestedAmount` in each redemption request is greater than
+ *          the sum of its `txFee` and `treasuryFee`
+ *        - the redeemer address in each redemption request is of a standard
+ *          type (P2PKH, P2WPKH, P2SH, P2WSH).
```

Comment:
Step 6: Let's remove this doc as it is no longer valid (remember about `index.ts`)

Reply:
Done in https://github.com/keep-network/tbtc-v2/pull/199/commits/4b3be0a5f176ab3e5733f1876ff60fa6d3ca53a2.